### PR TITLE
Support NHWC format in SpatialMaxPooling

### PR DESCRIPTION
**Pull request by yangw1234**
## What changes were proposed in this pull request?

Support NHWC format and SAME padding in SpatialMaxPooling (this pr needs to be merged after pr #1434 )

## How was this patch tested?

add unit tests

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialMaxPooling.scala [hunk 1]
```diff
@@ -69,169 +72,288 @@ class SpatialMaxPooling[T: ClassTag](
     this
   }
 
+  // return (padTop, padDown, padLeft, padRight)
+  protected def getPadding(inputHeight: Int, inputWidth: Int): (Int, Int, Int, Int) = {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialMaxPooling.scala]: I remember the same method appears in the conv2D layer. Can we make it as a utility function?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/NNPrimitive.scala [hunk 2]
```diff
@@ -721,6 +767,193 @@ private[nn] object NNPrimitive {
...
+    gradInputTensor: Tensor[Double],
+    gradOutputTensor: Tensor[Double],
+    indicesTensor: Tensor[Double],
+    oWidth: Int, oHeight: Int): Unit = {
+
+    val nSlices = gradInputTensor.size(3)
+
+    val gradInput = gradInputTensor.storage().array()
+    val gradInputOffset = gradInputTensor.storageOffset() - 1
+    val gradOutput = gradOutputTensor.storage()
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/NNPrimitive.scala]: why not pick the array?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/NNPrimitive.scala [hunk 3]
```diff
@@ -721,6 +767,193 @@ private[nn] object NNPrimitive {
...
+    val gradOutputOffset = gradOutputTensor.storageOffset() - 1
+    val indices = indicesTensor.storage()
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/NNPrimitive.scala]: why not pick the array?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/NNPrimitive.scala [hunk 4]
```diff
@@ -721,6 +767,193 @@ private[nn] object NNPrimitive {
...
+    indicesTensor: Tensor[Float],
+    oWidth: Int, oHeight: Int): Unit = {
+
+    val nSlices = gradInputTensor.size(3)
+
+    val gradInput = gradInputTensor.storage().array()
+    val gradInputOffset = gradInputTensor.storageOffset() - 1
+    val gradOutput = gradOutputTensor.storage()
+    val gradOutputOffset = gradOutputTensor.storageOffset() - 1
+    val indices = indicesTensor.storage()
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/NNPrimitive.scala]: why not pick the array?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/NNPrimitive.scala [hunk 5]
```diff
@@ -721,6 +767,193 @@ private[nn] object NNPrimitive {
...
+    gradInputTensor: Tensor[Float],
+    gradOutputTensor: Tensor[Float],
+    indicesTensor: Tensor[Float],
+    oWidth: Int, oHeight: Int): Unit = {
+
+    val nSlices = gradInputTensor.size(3)
+
+    val gradInput = gradInputTensor.storage().array()
+    val gradInputOffset = gradInputTensor.storageOffset() - 1
+    val gradOutput = gradOutputTensor.storage()
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/NNPrimitive.scala]: why not pick the array?

**Comment by yiheng** [comment]
Please also update the md docs

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialMaxPooling.scala [hunk 6]
```diff
@@ -69,169 +72,288 @@ class SpatialMaxPooling[T: ClassTag](
...
+    if (padW == -1 && padH == -1) {
+      // deal with SAME padding
+      val oW = Math.ceil(inputWidth.toFloat / dW.toFloat).toInt
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialMaxPooling.scala]: I remember pooling layer has a ceil option. Is this right for both cases?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/NNPrimitive.scala (hunk 2 above)
Comment (yangw1234) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/NNPrimitive.scala]: typo, I'll fix that

**Comment by yangw1234** [comment]
pr validation pass

**Comment by yiheng** [comment]
LGTM

**Comment by yangw1234** [comment]
pr validation pass

**Pull request merged**